File: public/econometrics/reproduce.R
# Reproduce the observed-worlds experiment in base R (no packages).
# Run: Rscript reproduce.R observed-worlds-seed-2026.csv
# With no argument, reads the fixed synthetic sample from the public site.
args <- commandArgs(trailingOnly=TRUE)
source <- if (length(args)) args[1] else "https://ihelfrich.github.io/econometrics/data/observed.csv"
d <- read.csv(source)
stopifnot(all(c("id", "x", "y") %in% names(d)), nrow(d) > 2)
print(coef(lm(y ~ x, data=d)))

# This latent decomposition is imposed as a model; it is not identified.
u <- d$x
v <- d$y-d$x
for (b in c(-1,0,.5,1,2)) {
  model.y <- b*d$x+(1-b)*u+v
  stopifnot(max(abs(model.y-d$y)) < 1e-12)
  cat("b =", b, ": same observed data; E[Y | do(X=1)] =", b, "\n")
}

dd <- (62-50)-(46-40)
M <- c(0,2,3,4,8)
result <- data.frame(M, effect.low=dd-M, effect.high=dd+M,
                     net.low=100*(dd-M)-300, net.high=100*(dd+M)-300)
print(result)
stopifnot(dd == 6, result$net.low[result$M==3] == 0)

# Optional second argument: CSV from the browser's sampling experiment.
if (length(args) >= 2) {
  s <- read.csv(args[2])
  width <- qnorm(.975)/sqrt(s$n)
  stopifnot(max(abs(s$low-(s$mean-width))) < 1e-12)
  stopifnot(max(abs(s$high-(s$mean+width))) < 1e-12)
  cat("Verified sampling coverage:", mean(s$low<=0 & s$high>=0), "\n")
}

File: public/econometrics/reproduce.py
"""Reproduce the observed-worlds experiment with the Python standard library.

Run: python reproduce.py [observed-worlds-seed-2026.csv]
Without an argument, downloads the public, fixed synthetic sample.
No real student or company data are used. No packages required.
Optional second argument: a sampling-seed-*.csv exported by the browser lab.
"""
import csv
import io
import math
import sys
import urllib.request
from pathlib import Path

if len(sys.argv) > 1:
    source = Path(sys.argv[1]).read_text(encoding="utf-8")
else:
    url = "https://ihelfrich.github.io/econometrics/data/observed.csv"
    with urllib.request.urlopen(url, timeout=20) as response:
        source = response.read().decode("utf-8")
rows = list(csv.DictReader(io.StringIO(source)))
x = [float(r["x"]) for r in rows]
y = [float(r["y"]) for r in rows]
n = len(x)
mx, my = sum(x) / n, sum(y) / n
slope = sum((a-mx)*(b-my) for a,b in zip(x,y)) / sum((a-mx)**2 for a in x)
intercept = my-slope*mx
print(f"Synthetic sample: n={n}; OLS intercept={intercept:.6f}, slope={slope:.6f}")

# U=X, V=Y-X reconstruct one latent model for demonstration; this is an
# assumed structural construction, not empirical discovery of confounders.
for b in [-1, 0, .5, 1, 2]:
    errors = [abs(yi-(b*xi+(1-b)*xi+(yi-xi))) for xi,yi in zip(x,y)]
    assert max(errors) < 1e-12
    print(f"b={b:4.1f}: same observed Y; E[Y | do(X=1)]={b:4.1f}")

# A separate algebraic sensitivity exercise: four means treated as known.
dd = (62-50)-(46-40)
print("\nM  effect bounds  net-value bounds ($)")
for bound in [0, 2, 3, 4, 8]:
    effects = (dd-bound, dd+bound)
    net = tuple(100*v-300 for v in effects)
    print(bound, effects, net)
assert dd == 6

# Export a batch from the sampling experiment, then supply it here to check
# all endpoints and the realized coverage using independent code.
if len(sys.argv) > 2:
    with Path(sys.argv[2]).open(newline="") as handle:
        simulations = list(csv.DictReader(handle))
    coverage = []
    for row in simulations:
        m, count = float(row["mean"]), int(row["n"])
        width = 1.959963984540054 / math.sqrt(count)
        lo, hi = m-width, m+width
        assert abs(lo-float(row["low"])) < 1e-12
        assert abs(hi-float(row["high"])) < 1e-12
        covered = lo <= 0 <= hi
        assert covered == (row["covers"].lower() == "true")
        coverage.append(covered)
    print(f"Verified {len(coverage)} intervals; coverage={sum(coverage)/len(coverage):.3f}")
